dataframe-arrow/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/ArrowWriterImpl.kt

```diff
@@ -0,0 +1,260 @@
+import org.apache.arrow.vector.Float8Vector
+import org.apache.arrow.vector.IntVector
+import org.apache.arrow.vector.LargeVarCharVector
+import org.apache.arrow.vector.SmallIntVector
+import org.apache.arrow.vector.TimeMicroVector
+import org.apache.arrow.vector.TimeMilliVector
+import org.apache.arrow.vector.TimeNanoVector
+import org.apache.arrow.vector.TimeSecVector
+import org.apache.arrow.vector.TinyIntVector
+import org.apache.arrow.vector.VarCharVector
+import org.apache.arrow.vector.VariableWidthVector
+import org.apache.arrow.vector.VectorSchemaRoot
+import org.apache.arrow.vector.types.DateUnit
+import org.apache.arrow.vector.types.FloatingPointPrecision
+import org.apache.arrow.vector.types.pojo.ArrowType
+import org.apache.arrow.vector.types.pojo.Field
+import org.apache.arrow.vector.types.pojo.FieldType
+import org.apache.arrow.vector.types.pojo.Schema
+import org.apache.arrow.vector.util.Text
+import org.jetbrains.kotlinx.dataframe.AnyCol
+import org.jetbrains.kotlinx.dataframe.DataFrame
+import org.jetbrains.kotlinx.dataframe.name
+import org.jetbrains.kotlinx.dataframe.api.convertToBigDecimal
+import org.jetbrains.kotlinx.dataframe.api.convertToBoolean
+import org.jetbrains.kotlinx.dataframe.api.convertToByte
+import org.jetbrains.kotlinx.dataframe.api.convertToDouble
+import org.jetbrains.kotlinx.dataframe.api.convertToFloat
+import org.jetbrains.kotlinx.dataframe.api.convertToInt
+import org.jetbrains.kotlinx.dataframe.api.convertToLocalDate
+import org.jetbrains.kotlinx.dataframe.api.convertToLocalDateTime
+import org.jetbrains.kotlinx.dataframe.api.convertToLocalTime
+import org.jetbrains.kotlinx.dataframe.api.convertToLong
+import org.jetbrains.kotlinx.dataframe.api.convertToShort
+import org.jetbrains.kotlinx.dataframe.api.convertToString
+import org.jetbrains.kotlinx.dataframe.api.forEachIndexed
+import org.jetbrains.kotlinx.dataframe.api.map
+import org.jetbrains.kotlinx.dataframe.exceptions.CellConversionException
+import org.jetbrains.kotlinx.dataframe.exceptions.TypeConverterNotFoundException
+
+/**
+ * Save [dataFrame] content in Apache Arrow format (can be written to File, ByteArray, OutputStream or raw Channel) with [targetSchema].
+ * If [dataFrame] content does not match with [targetSchema], behaviour is specified by [mode], mismatches would be sent to [mismatchSubscriber]
+ */
+internal class ArrowWriterImpl(
+    override val dataFrame: DataFrame<*>,
+    override val targetSchema: Schema,
+    override val mode: ArrowWriter.Companion.Mode,
+    override val mismatchSubscriber: (ConvertingMismatch) -> Unit = ignoreMismatchMessage
+) : ArrowWriter {
+
+    private val allocator = RootAllocator()
+
+    private fun allocateVector(vector: FieldVector, size: Int) {
+        when (vector) {
+            is FixedWidthVector -> vector.allocateNew(size)
+            is VariableWidthVector -> vector.allocateNew(size)
+            else -> TODO("Not implemented for ${vector.javaClass.canonicalName}")
```

Comment:
if you want to throw a NotImplementedError, please do so explicitly, not using a TODO(). But maybe an IllegalArgumentException might be better?